src/gensearch.jl

```diff
@@ -0,0 +1,336 @@
+"""
+# this is also minimal bd set!
+function find_min_covariate_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    PCPXY = pcp(g, X, Y)
+    return find_min_dsep(g, X, Y, I, setdiff(R, descendants(g, PCPXY)), (pe, ne, v, w) -> v in X && w in PCPXY && ne == RIGHT)
+end
+
+"""
+    find_frontdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+Find a frontdoor adjustment set `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, else return `false`. 
+
+Follows the algorithm given in  https://arxiv.org/abs/2211.16468. 
+"""
+function find_frontdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    Za = setdiff(R, bayesball(g, X, Set{Integer}(), no_outgoing(X)))
+    A = ancestors(g, Y)
+    function Zab_pass(pe, ne, v, w)
+        v in X && return false
+        ne == RIGHT && return true
+        pe == RIGHT && v in A && !(w in Za) && return true
+        (pe == INIT || pe == LEFT) && !(w in Za) && return true
+        return false
+    end
+    Zab = setdiff(Za, gensearch(g, Y, Zab_pass))
+    if issubset(I, Zab) && length(intersect(descendants(g, X, no_outgoing(Zab)), Y)) == 0
+        return Zab
+    else
+        return false
+    end
+end
+
+"""
+    find_min_frontdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+Find an inclusion minimal frontdoor adjustment set `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, else returns `false`. 
+
+Follows the algorithm given in  https://arxiv.org/abs/2211.16468. 
+"""
+# write explanation for the pass rules
+    function find_min_frontdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    Zii = find_frontdoor_adjustment(g, X, Y, I, R)
+    Zii == false && return false 
+    function Za_pass(pe, ne, v, w)
+        pe in [INIT, LEFT] && ne == LEFT && !(v in Zii) && !inunion(w, X, Y) && return true
+        return false
+    end
+    Za = intersect(gensearch(g, Y, Za_pass), Zii) 
+    function Zxy_pass(pe, ne, v, w)
+        pe in [INIT, RIGHT] && ne == RIGHT && !(v in Za) && !inunion(w, X, Y, I) && return true
+        return false
+    end
+    Zxy = intersect(gensearch(g, X, Zxy_pass), Za)
+    function Zzy_pass(pe, ne, v, w)
+        pe in [INIT, LEFT] && ne == LEFT && !inunion(w, X, I, Zxy) && return true
+        pe in [LEFT, RIGHT] && ne == RIGHT && !(w in X) && !inunion(v, I, Za) && return true
+        pe == RIGHT && ne == LEFT && inunion(v, I, Za) && !inunion(w, X, I, Zxy) && return true
+        return false
+    end
+    Zzy = intersect(gensearch(g, union(I, Zxy), Zzy_pass), Za) 
+    return union(I, Zxy, Zzy)
+end
+
+struct ConstraintIterator{T<:Integer, F<:Function}
+    g::SimpleDiGraph{T}
+    X::Set{T}
+    Y::Set{T}
+    I::Set{T}
+    R::Set{T}
+    find::F
+end
+
+function downwards(state, I, R)
+    v = first(setdiff(R, I))
+    push!(state, ("up", "I", v, I, R))
```

Comment:
replace "strings" with :symbols

Reply:
done!

Potentially one can write this part more elegantly by accessing the sets I and R via symbols, but for now I think it's fine